### ISSUE TITLE
"Header Checksum" in ipv4 header - what is the relevance of pattern here ?

### ISSUE THREAD

**Issue by indraniBh** [open-traffic-generator/models#31]
Should not the values (enum) 'valid checksum' and 'invalid checksum' suffice instead of pattern ?

**Comment by ashutshkumr** [comment]
@ajbalogh we should probably get rid of the pattern for any fields that require checksum. Reasoning is, checksums are usually calculated and embedded right before egress. So whatever we put there might be overwritten.

Although, we might want to allow user to state whether they want to embed bad checksum or good checksum.

**Comment by ankur-sheth** [comment]
Agree.  @ajbalogh thoughts?

**Comment by winstonliu1111** [comment]
@indraniBh @ashutshkumr do you guys want to make a proposal on what the model would look like?  (what would go in place of the pattern concept, that can be multi-purposed to various checksum/crc needs for potentially different packet headers that have those type of needs)?

**Comment by ashutshkumr** [comment]
We do not have a proposal, just wanted to highlight the problem. Most probably we'll need a checksum container to be assigned to all possible checksum fields in l2-4.

e.g.
```python
# automatically insert correct checksum
h = FlowHeader(choice='ipv4', ipv4=Ipv4())
h = FlowHeader(choice='ipv4', ipv4=Ipv4(header_checksum=Checksum(choice='good')))

# automatically insert incorrect checksum
h = FlowHeader(choice='ipv4', ipv4=Ipv4(header_checksum=Checksum(choice='bad')))

# 16-bit checksum
h = FlowHeader(choice='ipv4', ipv4=Ipv4(header_checksum=Checksum(choice='custom', custom='aabb')))
# 32-bit checksum
h = FlowHeader(choice='ethernet', ethernet=Ethernet(fcs=Checksum(choice='custom', custom='aabbccdd')))
```

**NOTE:** We can only provide fixed value for custom checksum (ask traffic engine to insert specific checksum before egress).

**Comment by ajbalogh** [comment]
I don't think it needs a different container as its just another packet field that can be standard or non-standard.
As a user if I'm not interested in the checksum I'm letting the system do it for me - the system's version of a checksum.
If I'm interested in setting the checksum I would like to set it to a single value, value list, counter etc. which will result in either a good or bad checksum.

**Comment by ashutshkumr** [comment]
If one wants to insert bad checksum, they shouldn't rely on providing custom checksum (which requires making sure it's incorrect). 
Because, we cannot guarantee that packet contents won't be altered (most probable cause is embedding instrumentation).

Hence, it's almost always easier to just request for bad checksum insertion, which we cannot do with `Pattern`.

As for custom checksum insertion, I don't have anything against multiple values. If it's not supported by traffic engine, we can always raise it as error.

**Comment by ajbalogh** [comment]
x-field-pattern extension allows for specifying `format: checksum` which generates a choice defaulted to an auto generated system checksum with the option for a user to specify their own custom checksum.  Available in release 0.3.1
